File: TextosJapones/src/main/java/org/jisho/textosJapones/database/dao/DaoFactory.java
package org.jisho.textosJapones.database.dao;

import org.jisho.textosJapones.database.dao.implement.*;
import org.jisho.textosJapones.database.mysql.DB;
import org.jisho.textosJapones.model.entities.VocabularioExterno;
import org.jisho.textosJapones.model.enums.Conexao;

import java.sql.Connection;
import java.util.ArrayList;
import java.util.List;

public class DaoFactory {

	public static VocabularioDao createVocabularioJaponesDao() {
		return new VocabularioJaponesDaoJDBC(DB.getLocalConnection());
	}
	
	public static VocabularioDao createVocabularioInglesDao() {
		return new VocabularioInglesDaoJDBC(DB.getConnection(Conexao.TEXTOINGLES));
	}

	public static EstatisticaDao createEstatisticaDao() {
		return new EstatisticaDaoJDBC(DB.getLocalConnection());
	}

	public static RevisarDao createRevisarJaponesDao() {
		return new RevisarJaponesDaoJDBC(DB.getLocalConnection());
	}
	
	public static RevisarDao createRevisarInglesDao() {
		return new RevisarInglesDaoJDBC(DB.getConnection(Conexao.TEXTOINGLES));
	}

	public static LegendasDao createLegendasDao() {
		return new LegendasDaoJDBC(DB.getLocalConnection(), DB.getConnection(Conexao.DECKSUBTITLE), DB.getDados(Conexao.DECKSUBTITLE).getBase());
	}

	public static MangaDao createMangaDao() {
		return new MangaDaoJDBC(DB.getConnection(Conexao.MANGAEXTRACTOR), DB.getDados(Conexao.MANGAEXTRACTOR).getBase());
	}

	public static NovelDao createNovelDao() {
		return new NovelDaoJDBC(DB.getConnection(Conexao.NOVELEXTRACTOR), DB.getDados(Conexao.NOVELEXTRACTOR).getBase());
	}
	
	public static VincularDao createVincularDao() {
		return new VincularDaoJDBC(DB.getConnection(Conexao.MANGAEXTRACTOR));
	}

	public static KanjiDao createKanjiDao() {
		return new KanjiDaoJDBC(DB.getLocalConnection());
	}

	public static ComicInfoDao createComicInfoDao() {
		return new ComicInfoJDBC(DB.getLocalConnection());
	}

	public static List<VocabularioDao> getVocabularioExternos() {
		List<VocabularioDao> externos = new ArrayList<>();

		Connection extractor = DB.getConnection(Conexao.MANGAEXTRACTOR);
		if (extractor != null)
			externos.add(new VocabularioExternoDaoJDBC(extractor));

		Connection novel = DB.getConnection(Conexao.NOVELEXTRACTOR);
		if (novel != null)
			externos.add(new VocabularioExternoDaoJDBC(novel));

		return externos;
	}

}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/database/DaoFactory.kt
package br.com.fenix.processatexto.database

import br.com.fenix.processatexto.database.dao.*
import br.com.fenix.processatexto.database.dao.implement.*
import br.com.fenix.processatexto.model.enums.Conexao
import org.slf4j.Logger
import org.slf4j.LoggerFactory


object DaoFactory {

    private val LOGGER: Logger = LoggerFactory.getLogger(DaoFactory::class.java)

    fun createVocabularioJaponesDao(): VocabularioDao = VocabularioJaponesDaoJDBC(Conexao.TEXTO_JAPONES)

    fun createVocabularioInglesDao(): VocabularioDao = VocabularioInglesDaoJDBC(Conexao.TEXTO_INGLES)

    fun createEstatisticaDao(): EstatisticaDao = EstatisticaDaoJDBC(Conexao.TEXTO_JAPONES)

    fun createRevisarJaponesDao(): RevisarDao = RevisarJaponesDaoJDBC(Conexao.TEXTO_JAPONES)

    fun createRevisarInglesDao(): RevisarDao = RevisarInglesDaoJDBC(Conexao.TEXTO_INGLES)

    fun createLegendasDao(): LegendasDao? {
        val dados = JdbcFactory.getConfiguracao(Conexao.DECKSUBTITLE)
        return if (dados.isPresent) LegendasDaoJDBC(Conexao.DECKSUBTITLE, dados.get().base) else null
    }

    fun createMangaDao(): MangaDao? {
        val dados = JdbcFactory.getConfiguracao(Conexao.MANGA_EXTRACTOR)
        return if (dados.isPresent) MangaDaoJDBC(Conexao.MANGA_EXTRACTOR, dados.get().base) else null
    }

    fun createNovelDao(): NovelDao? {
        val dados = JdbcFactory.getConfiguracao(Conexao.NOVEL_EXTRACTOR)
        return if (dados.isPresent) NovelDaoJDBC(Conexao.NOVEL_EXTRACTOR, dados.get().base) else null
    }

    fun createVincularDao(): VincularDao? {
        val dados = JdbcFactory.getConfiguracao(Conexao.NOVEL_EXTRACTOR)
        return if (dados.isPresent) VincularDaoJDBC(Conexao.NOVEL_EXTRACTOR) else null
    }

    fun createKanjiDao(): KanjiDao = KanjiDaoJDBC(Conexao.TEXTO_JAPONES)

    fun createComicInfoDao(): ComicInfoDao = ComicInfoJDBC(Conexao.PROCESSA_TEXTO)

    fun createSincronizacaoDao(): SincronizacaoDao = SincronizacaoDaoJDBC(Conexao.PROCESSA_TEXTO)

    val vocabularioExternos: List<VocabularioDao>
        get() {
            val externos: MutableList<VocabularioDao> = mutableListOf()

            if (JdbcFactory.getConfiguracao(Conexao.MANGA_EXTRACTOR).isPresent)
                externos.add(VocabularioExternoDaoJDBC(Conexao.MANGA_EXTRACTOR))

            if (JdbcFactory.getConfiguracao(Conexao.NOVEL_EXTRACTOR).isPresent)
                externos.add(VocabularioExternoDaoJDBC(Conexao.NOVEL_EXTRACTOR))

            return externos
        }
}